### ISSUE TITLE
scene.background support for equirectangular textures

### ISSUE THREAD

**Issue by mkeblx** [mrdoob/three.js#9733]
Currently the helper `background` property on a `Scene` can be set to either a cubemap (rendered on a cube geometry) or a texture on a fullscreen orthographic plane. There are a lot of equirectangular images (a lot more coming from 360/VR cameras soon) might be nice to also support as another type and get same benefits.

Potential API:

``` js
scene.background = new THREE.TextureLoader().load( "textures/360-cam.jpg" );
scene.backgroundSphere = true; 

// OR : more flexible can define mesh
scene.background = new THREE.TextureLoader().load( "textures/equirect.jpg" );
scene.backgroundMesh = new THREE.Mesh(..., ...);
```

**Comment by mrdoob** [comment]
How aboud doing a Equirectangular to Cubemap helper instead?

/ping @spite

**Comment by spite** [comment]
I've got half working a helper to convert equirectangular panoramas to THREE.WebGLRenderTargetCube (based on https://github.com/spite/THREE.CubemapToEquirectangular)

It makes more sense to use cube mapping rather than spherical mapping, since there's cases in which the seams in the wrapping can't be fixed.

**Comment by spite** [comment]
Pick your poison:
- https://github.com/spite/THREE.EquirectangularToCubemap
- https://github.com/mrdoob/three.js/pull/9746 

🍰🍰🍰

**Comment by mkeblx** [comment]
Awesome thanks @spite. These should be sufficient. I did want to check one thing with #9746.

**Comment by mrdoob** [comment]
I'm thinking we may want to add `THREE.EquirectangularTexture()`.

**Comment by makc** [comment]
> It makes more sense to use cube mapping rather than spherical mapping, since there's cases in which the seams in the wrapping can't be fixed.

I had someone telling me that they get seams in cubemaps, https://github.com/KhronosGroup/WebGL/issues/1528 this would not happen with the equirectangular image, although you need a bit heavier shader

**Comment by spite** [comment]
@makc https://github.com/mrdoob/three.js/issues/1621

**Comment by makc** [comment]
@spite check this [4 tap trick](https://0fps.net/2013/07/09/texture-atlases-wrapping-and-mip-mapping/) article

**Comment by spite** [comment]
That will make capturing trickier, since the CSS background won't be part of the Canvas. It might be better to have a backdrop/background object that can work with equi/cubemap/flat textures (sphere/cube/tri geos)?

**Comment by WestLangley** [comment]
>That will make capturing trickier, since the CSS background won't be part of the Canvas.

@spite Good point.

Maybe an API that captures these features...
```javascript
renderer.clearColor
renderer.backgroundTexture // image or procedural texture, backdrop
scene.background //equirectangular or cube
```

**Comment by mrdoob** [comment]
@WestLangley We could also add new mappings... `if( scene.background.isTexture && scene.background.mapping === EquirectangularMapping )`.

**Comment by WestLangley** [comment]
@mrdoob Yes, I think that is a good idea. We would have to make sure it still works if the user sets the mapping in the loader callback -- and hence, potentially after the first call to render.

**Comment by DavidPeicho** [comment]
Has the mapping been abandoned in favor of the PMREM conversion?

I imagine having the mapping directly done in the shader is still nice for the background. What do you think? @mrdoob @WestLangley

**Comment by mrdoob** [comment]
> I imagine having the mapping directly done in the shader is still nice for the background. What do you think?

I considered this...

Something like a `THREE.EquirectangularTexture` which the renderer could internally do a `PMREMGenerator` pass when used as `scene.environment`.

🤔

/cc @elalish

**Comment by DavidPeicho** [comment]
Do we need to use the `PMREMGenerator`, or should we just handle a new mapping like you said?
```javascript
if( scene.background.isTexture && scene.background.mapping === EquirectangularMapping ) {

        // Set shader define to `ENVMAP_TYPE_EQUIREC `

}
```
And fetch the texture using spherical coordinates in the shader.

**Comment by WestLangley** [comment]
@DavidPeicho
```js
var options = {
	generateMipmaps: true,
	minFilter: THREE.LinearMipmapLinearFilter,
	magFilter: THREE.LinearFilter
};

scene.background = new THREE.WebGLCubeRenderTarget( 512, options ).fromEquirectangularTexture( renderer, texture );
```

**Comment by DavidPeicho** [comment]
@WestLangley Interesting! I missed that, really nice and done on the GPU.

It would still be nice to support equirectangular like any other texture as a background with no conversion, but it's not a priority.

**Comment by marcofugaro** [comment]
thank you @WestLangley, that was useful. 

How should the WebGLCubeRenderTarget size be chosen in order to get the best resolution from the image?

I am doing something like [this example](https://threejs.org/examples/webgl_panorama_equirectangular), with a really hi-res equirectangular image.
Right now I'm getting the closest power of two from the height, but I agree that this should be done in a better way.

```js
const size = highestPowerOfTwo(texture.image.naturalHeight)
```

**Comment by elalish** [comment]
@marcofugaro I generally use half the height (assuming powers of two). So for a 1024x512 equirect, I use a 256x256 cubemap. That gives identical pixel sizes at the equator, and the cube map gets 25% fewer pixels overall, since the stretched poles get collapsed.

**Comment by Mugen87** [comment]
Fixed via #19911 🎉
